<span style='color:brown'>Python es un lenguaje de programación ampliamente útil y orientado a objetos. Su sentido significativo junto a sus increíbles bibliotecas le han dado el respeto de ser el idioma preferido para carreras como la de un científico de datos o un experto en IA.</span>

<span style='color:brown'>Python también es elegido con frecuencia como el lenguaje para la programación en escuelas y universidades debido a su facilidad de aprendizaje. Estos tutoriales instructivos sirven para ayudar a aprender Python desde cero y hasta un nivel avanzado en este lenguaje.</span>

<a id='primeros_pasos'></a>
## Lección 1 - Introducción a la programación con Python

#### Primeros pasos con Python y Jupyter

![Operadores](Python_operators.png)

### Parte 1 de "Análisis de datos con Python

Esta serie de tutoriales es una introducción amigable para principiantes a la programación y el análisis de datos utilizando el lenguaje de programación python. Estos tutoriales toman un enfoque práctico y centrado en la codificación. La mejor manera de aprender el material es ejecutar el código y experimentar con él usted mismo. Echa un vistazo a la serie completa aquí:

>1. [Primeros pasos con Python y Jupyter](#primeros_pasos)
>2. [Un recorrido rápido por variables y tipos de datos](#variables)
>3. Bifurcación mediante instrucciones y bucles condicionales
>4. Escribir código reutilizable mediante funciones
>5. Leer y escribir a archivos
>6. Computación numérica con Python y Numpy
>7. Análisis de datos tabulares con pandas
>8. Visualización de datos mediante Matplotlib & Seaborn
>9. Análisis exploratorio de datos - Un caso de estudio

En este tutorial se tratan los siguientes temas:

>* Realización de operaciones aritméticas con Python
>* Solución de problemas de varios pasos mediante variables
>* Evaluación de condiciones mediante Python
>* Combinación de condiciones con operadores lógicos
>* Adición de estilos de texto mediante Markdown

### Realización de operaciones aritméticas con Python
Empecemos por usar Python como una calculadora. Puedes escribir y ejecutar Python mediante una celda del tipo `código` dentro de tu cuaderno de Jupyter.

<ins>Trabajar con celdas</ins>: Para crear una nueva celda dentro de Jupyter, puedes seleccionar "Insertar > Insertar celda a continuación" en la barra de menús o simplemente presione la tecla <kbd>+</kbd> en la barra de herramientas.
También puedes usar el método abreviado de teclado para crear una nueva celda (presionando la tecla <kbd>a</kbd> para insertar una celda anterior a la seleccionada o <kbd>b</kbd> para insertar una celda después). Una vez creada una celda, haga clic en la misma para seleccionarla, a continuación, puede cambiar el tipo de celda a código o markdown (texto) mediante la opción de menú "Celda > Tipo de celda".
Para editar el contenido la celda haga doble clic en la misma.
Para aplicar los cambios y ejecutar una celda, utilice la opción de menú "Celda > Ejecutar celdas" o haga clic en el botón "Ejecutar" de la barra de herramientas o simplemente utilice el método abreviado de teclado (<kbd>Ctrl</kbd>+<kbd>Intro</kbd> o <kbd>Shift</kbd>+<kbd>Intro</kbd>).
Puedes ver una lista completa de métodos abreviados de teclado mediante la opción de menú "Ayuda > Métodos abreviados de teclado".

### Realización de operaciones aritméticas con Python
Comencemos por usar Python como una calculadora. Puedes escribir y ejecutar Python mediante una celda de código dentro de Jupyter.

Ejecute las celdas de código siguientes para realizar cálculos y ver su resultado. Intente cambiar los números y vuelva a ejecutar las celdas modificadas para ver los resultados actualizados. ¿Puedes adivinar para qué se utilizan los operadores `// % **`?

<span style="color:brown">_Por cierto, utilizamos el símbolo "#" para inserir comentarios en nuestras celdas de código_</span>

In [1]:
# Esto es un comentario, no será ejecutado en nuestro código

In [2]:
# Suma
2 + 3 + 9

14

In [3]:
# Resta
99 -73

26

In [4]:
# Multiplicación
23.54 * -1432

-33709.28

In [5]:
# División
100 / 7

14.285714285714286

In [6]:
# Parte entera de la División
100 // 7

14

In [7]:
# Resto de la división
100 % 7

2

In [8]:
# Potenciación
5 ** 3

125

Como era de esperar, los operadores tienen prioridad sobre otros operadores como y según convenciones matemáticas. Puedes utilizar paréntesis para separar y ordenar cómo se realizan las operaciones. `( ) ** * / % // + -`

In [9]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

#### Solución de problemas de varios pasos mediante variables

Intentemos resolver el siguiente problema:

Una tienda de comestibles vende una bolsa de hielo por $1.25 y obtiene un beneficio del 20%. Si vende 500 bolsas de hielo, ¿cuánto beneficio total obtienen?

Podemos enumerar la información proporcionada y convertir gradualmente la palabra problema en una expresión matemática que se puede evaluar mediante Python.

Coste de la bolsa de hielo ($) = 1.25

Margen de beneficio = 20% = .2

Beneficio por bolsa ($) = margen de beneficio * costo de la bolsa de hielo = .2 * 1.25

Número de bolsas = 500

Beneficio total = número de bolsas * beneficio por bolsa = 500 * (.2 * 1.25)

In [10]:
500 * (.2 * 1.25)

125.0

Por lo tanto, la tienda de comestibles obtiene un beneficio total de $125. Si bien esta es una manera razonable de resolver un problema, no está del todo claro mirando la celda de código lo que representan los números. Podemos dar nombres a cada uno de los números mediante la creación de variablesde Python.

Mientras se trabaja con un lenguaje de programación como Python, la información se almacena en **variables**. Puede pensar en variables como contenedores para almacenar datos. Los datos almacenados dentro de una variable se denominan su valor.

In [11]:
# Declaremos las variables
coste_bolsa_hielo = 1.25
margen_beneficio = .2
num_bolsas = 500

Las variables ahora contienen los valores proporcionados. Podemos comprobar el valor de una variable escribiendo su nombre en una celda. También podemos combinar variables mediante operaciones aritméticas para crear otras variables.

Pista: al escribir el nombre de una variable existente en una celda de código dentro de Jupyter, simplemente escriba los primeros caracteres y presione la tecla "Tab" para autocompletar el nombre de la variable. Intente escribir una celda de código a continuación y pulse para autocompletarla.

In [12]:
coste_bolsa_hielo

1.25

In [13]:
margen_beneficio

0.2

In [14]:
num_bolsas

500

In [15]:
# Creamos/declaramos la variable margen_benef_por_bolsa
margen_benef_por_bolsa = margen_beneficio * coste_bolsa_hielo

In [16]:
margen_benef_por_bolsa

0.25

In [17]:
# Creamos/declaramos la variable margen_total
margen_total = num_bolsas * margen_benef_por_bolsa

In [18]:
margen_total

125.0

Si intentas ver el valor de una variable que no se ha definido antes, es decir, dado un valor mediante la instrucción, Python mostrará un error (name 'nombre_variable' is not defined)

In [19]:
Margen_neto

NameError: name 'Margen_neto' is not defined

Almacenar y manipular datos utilizando variables con el nombre adecuado es una excelente manera de explicar lo que hace el código.

Función `PRINT`: Para exibir la información o resultado de una celda utilizamos la función `print`.

Funciones: Una función es un conjunto reutilizable de instrucciones. Toma una o más entradas, realiza ciertas operaciones y a menudo devuelve una salida. Python proporciona muchas funciones integradas, pero también nos permite definir nuestras propias funciones.

In [20]:
print("La tienda de comestibles obtuvo un beneficio total de $", margen_total)

La tienda de comestibles obtuvo un beneficio total de $ 125.0


Función **PRINT**: La función `PRINT` se utiliza para mostrar información. Se necesitan una o más entradas, que pueden ser texto (entre comillas simples `'` o duplas `"`), números, variables, expresiones matemáticas, etc. Veremos más adelante sobre variables y funciones en el siguiente tutorial.

Crear una celda de código para cada operación de variable o operación matemática puede volverse tedioso. Afortunadamente, Jupyter nos permite escribir varias líneas de código dentro de una sola celda de código y ejecutarlos todos de una sola vez.

Reescribamos la solución a nuestro problema de palabras dentro de una sola celda.

In [21]:
# Declaremos las variables
coste_bolsa_hielo = 1.25
margen_beneficio = .2
num_bolsas = 500

# Declaremos las variables de resultado realizando los cálculos con las variables anteriores
margen_benef_por_bolsa = margen_beneficio * coste_bolsa_hielo
margen_total = num_bolsas * margen_benef_por_bolsa

# Mostremos en pantalla el resultado
print("La tienda de comestibles obtuvo un beneficio total de $", margen_total)

La tienda de comestibles obtuvo un beneficio total de $ 125.0


Recuerda que estamos usando el carácter # para agregar comentarios dentro de nuestro código.

**Comentarios**: Los comentarios y las líneas en blanco se ignoran durante la ejecución, pero son útiles para proporcionar información a los seres humanos (incluido usted mismo) sobre lo que hace el código. Los comentarios pueden estar en línea (al final de algún código), en una línea independiente o incluso abarcar varias líneas.

Los comentarios en línea y de una sola línea comienzan con `#`, mientras que los comentarios multilínea comienzan y terminan con tres comillas. Ejemplo: '''Este es un cometario de varias lineas'''.

In [22]:
mi_num_pref = 9 # Este es un comentario de ua sola línea

In [23]:
# Este también es un comentario de una sola línea
mi_peor_num = 1

"""Este es un comentario de varias líneas.
Escribe lo poco o lo mucho que quieras.

Los comentarios son realmente útiles para la gente que lee tu código, pero trate de mantenerlos cortos y al punto.

Además, si usas buenos nombres de variables, tu código suele ser un código que se explica por sí mismo, ¡y puede que ni siquiera necesites comentarios!
"""

a_número_neutral = 5

#### Evaluación de condiciones mediante Python
Además de las operaciones aritméticas, Python también proporciona varias operaciones para comparar números y variables.

![Operadores_comparacion](Operadores_comparacion.png)

El resultado de una operación de comparación será siempre `True` o `False` (tenga en cuenta las mayúsculas T y F). Estas son palabras clave especiales en Python. Probemos algunos experimentos con operadores de comparación.

In [24]:
mi_num_fav = 1
mi_ultim_num_fav = 5
mi_num_neutral = 3

In [25]:
# Operación Igual que - True
mi_num_fav == 1

True

In [26]:
# Operación Igual que - False
mi_num_fav == mi_ultim_num_fav

False

In [27]:
# Operación de Distinto que - True
mi_num_fav != mi_num_neutral

True

In [28]:
# Operación de Distinto que - False
mi_num_neutral != 3

False

In [29]:
# Operación de Mayor que - True
mi_ultim_num_fav > mi_num_neutral

True

In [30]:
# Operación de Mayor que - False
mi_num_fav > mi_ultim_num_fav

False

In [31]:
# Operación Menor que - True
mi_num_fav < 10

True

In [32]:
# Operación Menor que - False
mi_ultim_num_fav < mi_num_fav

False

In [33]:
# Operación Mayor o igual que - True
mi_num_fav >= 1

True

In [34]:
# Operación Mayor o igual que - False
mi_num_fav >= 3

False

In [35]:
# Operación Menor o igual que - True
3 + 6 <= 9

True

In [36]:
# Operación Menor o igual que - False
mi_num_fav + mi_num_neutral <= 3

False

Al igual que las operaciones aritméticas, el resultado de una operación de comparación también se puede almacenar en una variable.

In [37]:
coste_bolsa_hielo = 1.25
es_la_bolsa_de_hielo_cara = coste_bolsa_hielo >= 10
print("Está la bolsa de hielo cara?", es_la_bolsa_de_hielo_cara)

Está la bolsa de hielo cara? False


#### Combinación de condiciones con operadores lógicos
Los operadores lógicos `and`, `or` y `not` operan según condiciones y valores `True` y `False` (también conocidos como booleanos). `and` y `or` operan en dos condiciones, mientras que `not` opera en una sola condición.

El operador retornará `True` cuando ambas condiciones se evalúan como `True`. De lo contrario, retornará `False`.

![Operadores_booleanos](Operadores_booleanos.jpg)

In [38]:
mi_num_neutral = 3

In [39]:
mi_num_neutral == 3 or mi_num_fav > 0

True

In [40]:
mi_num_neutral != 3 or mi_num_fav > 0

True

In [41]:
mi_num_fav < 0 or True

True

In [42]:
False or False

False

El operador `not` devuelve `False` si una condición es `True` y `True` si la condición es `False`.

In [43]:
not mi_num_neutral == 3

False

In [44]:
not mi_num_fav < 0

True

In [45]:
not False

True

In [46]:
not True

False

Los operadores lógicos se pueden combinar para formar condiciones complejas. Utilice corchetes o paréntesis para indicar el orden en que se deben aplicar los operadores lógicos.

In [47]:
(2 > 3 and 4 <= 5) or not (mi_num_fav < 0 and True)

True

In [48]:
not (True and 0 < 1) or (False and True)

False

Si no se utilizan paréntesis, los operadores lógicos se aplican de izquierda a derecha.

In [49]:
not True and 0 < 1 or False and True

False

Experimentemos con operadores aritméticos, condicionales y lógicos en Python utilizando la naturaleza interactiva de los cuadernos jupyter. Aprenderemos más sobre variables y funciones más adelante.

#### Adición de estilos de texto mediante Markdown
Agregar explicaciones usando celdas de texto (como esta) es una gran manera de hacer que tu cuaderno sea informativo para otros lectores. También es útil si necesitas volver a él en el futuro. Puede hacer doble clic en una celda de texto dentro de Jupyter para editarla. En el modo de edición, observará que el texto se ve ligeramente diferente (por ejemplo, el encabezado tiene un `##` como prefijo). Este texto se escribe con Markdown, una forma sencilla de agregar estilos al texto. Ejecute esta celda para ver la salida sin los caracteres especiales. Puede cambiar de un lado a otro entre la fuente y la salida para aplicar un estilo específico.

Por ejemplo, puedes utilizar uno o varios `#` al inicio de una línea para crear encabezados de diferentes tamaños:

# Encabezado con un `#`
## Encabezado con dos `#`
### Encabezado con tres `#`
#### Encabezado con cuatro `#`
##### Encabezado con cinco `#`
###### Encabezado con seis `#`


Para crear una lista con viñetas o numerada, simplemente inicie una línea con `*` o `1`.

Una lista con viñetas:
* Punto 1
* Punto 2
* Punto 3

Una lista numerada:
1. manzana
2. Plátano
3. Piña

Puedes insertar texto en negrita utilizando doble asterisco, por ejemplo, **algún texto en negrita**, o hacerlo cursivo utilizando un solo asterisco, por ejemplo, *algún texto cursivo*. También puede crear enlaces, insertar imágenes, etc..

Para más información acerca de todo lo que se puede hacer con Mardowns puedes visitar:
[https://learnxinyminutes.com/docs/markdown/](]https://learnxinyminutes.com/docs/markdown/)

<a id='variables'></a>
### Un recorrido rápido por variables y tipos de datos

![Variables](variables_data_types.png)

### Parte 2 de "Análisis de datos con Python

>1. [Primeros pasos con Python y Jupyter](#primeros_pasos)
>2. [Un recorrido rápido por variables y tipos de datos](#variables)
>3. Bifurcación mediante instrucciones y bucles condicionales
>4. Escribir código reutilizable mediante funciones
>5. Leer y escribir a archivos
>6. Computación numérica con Python y Numpy
>7. Análisis de datos tabulares con pandas
>8. Visualización de datos mediante Matplotlib & Seaborn
>9. Análisis exploratorio de datos - Un caso de estudio
En este tutorial se tratan los siguientes temas:

En este tutorial se tratan los siguientes temas:

>* Almacenamiento de información mediante variables
>* Tipos de datos primitivos en Python: Entero, Flotante, Booleano, Ninguno y Cadena
>* Estructuras de datos integradas en Python: Lista, Tupla y Diccionario
>* Métodos y operadores compatibles con tipos de datos integrados

### Almacenamiento de información mediante variables
Las variables son útiles para dos propósitos: almacenar información (también conocida como datos) y realizar operaciones en datos almacenados. Mientras se trabaja con un lenguaje de programación como Python, los datos se almacenan en variables. Puedes pensar en variables como contenedores para almacenar datos y donde los datos almacenados dentro de una variable se denominan su valor. Crear variables en Python es bastante fácil, como ya hemos visto anteriormente.

In [50]:
mi_color_favorito = 'azul'

In [51]:
mi_color_favorito

'azul'

Se crea una variable mediante una instrucción de asignación `=`. Comienza con el nombre de la variable, seguido por el operador de asignación `=` seguido del valor que se almacenará dentro de la variable. Tenga en cuenta que el operador de asignación `=` es diferente del operador de comparación de igualdad `==`.

También puedes asignar valores a varias variables de una sola instrucción separando los nombres y valores de las variables con comas.

In [52]:
color1, color2, color3 = 'azul', 'rojo', 'verde'

In [53]:
color1

'azul'

In [54]:
color2

'rojo'

In [55]:
color3

'verde'

También puedes asignar el mismo valor a varias variables encadenando varias operaciones de asignación dentro de una sola instrucción.

In [56]:
color4 = color5 = color6 = 'morado'

In [57]:
color4

'morado'

In [58]:
color5

'morado'

In [59]:
color6

'morado'

Puedes cambiar el valor almacenado dentro de una variable asignándole un nuevo valor mediante otra instrucción de asignación. Tenga cuidado al reasignar variables: al asignar un nuevo valor a la variable, el valor antiguo se pierde y ya no es accesible.

In [60]:
mi_color_favorito = 'rojo'

In [61]:
mi_color_favorito

'rojo'

Al reasignar una variable, también puede usar el valor anterior de la variable para calcular el nuevo valor.

In [62]:
contador = 10

In [63]:
contador = contador + 1

In [64]:
contador

11

El patrón (donde hay un operador aritmético como `+`, `-`, `*`, `/`) es muy común, por lo que Python proporciona una sintaxis abreviada para él.

In [65]:
contador = 10

In [66]:
# Igualmente que contador = contador + 4
contador += 4

In [67]:
contador

14

Los nombres de variables pueden ser cortos (a, b, x, etc...) o descriptivos ( mi_color_favorito, margen_beneficio, los_3_mosqueteros, etc...). Sin embargo, debes seguir estas reglas al nombrar variables de Python:

* El nombre de una variable debe empezar con una letra o el carácter de subrayado `_`. No debe empezar por un número
* Un nombre de variable solo puede contener letras mayúsculas o minúsculas (mayúsculas), dígitos o guiones bajos
* Los nombres de variables se distinguen entre mayúsculas y minúsculas, es decir, son diferentes dependiendo de cómo se cree.

Estos son algunos nombres de variables válidos:

In [68]:
a_variable = 23
hoy_es_Sabado = False
mi_coche_favorito = "Aston Martin"
los_3_mosqueteros = ['Athos', 'Porthos', 'Aramis']

Intentemos crear algunas variables con nombres no válidos. Python imprime un error de sintaxis si el nombre de la variable no es válido.

> Sintaxis: La sintaxis de un lenguaje de programación hace referencia a las reglas que rigen la estructura de una instrucción o instrucción válida. Si una instrucción no sigue estas reglas, Python detiene la ejecución e le informa de que hay un error de sintaxis. Puede pensar en la sintaxis como las reglas de gramática para un lenguaje de programación.

In [69]:
a variable = 23

SyntaxError: invalid syntax (<ipython-input-69-b4f9daa715a5>, line 1)

In [ ]:
hoy_es_$abado = False

In [ ]:
mi-coche-favorito = 'Aston Martin'

In [ ]:
3_mosqueteros = ['Athos', 'Porthos', 'Aramis']

### Tipos de datos integrados en Python
Cualquier dato o información almacenado dentro de una variable de Python tiene un tipo. Puede ver el tipo de datos almacenados dentro de una variable mediante la función `type`.

In [70]:
a_variable

23

In [71]:
type(a_variable)

int

In [72]:
hoy_es_Sabado

False

In [73]:
type(hoy_es_Sabado)

bool

In [74]:
mi_coche_favorito

'Aston Martin'

In [75]:
type(mi_coche_favorito)

str

In [76]:
los_3_mosqueteros

['Athos', 'Porthos', 'Aramis']

In [77]:
type(los_3_mosqueteros)

list

Python tiene varios tipos de datos integrados para almacenar diferentes tipos de información en variables. A continuación se muestran algunos tipos de datos de uso común:

Entero (int)  
Flotante/decimal (float)  
Booleana (bool)  
Ninguno (None)  
Cadena (str)  
Lista (list)  
Tupla (tuple)  
Diccionario (dictionary)  

Integer, float, boolean, None y string son tipos de datos primitivos porque representan un único valor. Otros tipos de datos como list, tuple y dictionary a menudo se denominan estructuras de datos o contenedores porque mantienen varios fragmentos de datos juntos.

#### Entero
Los enteros representan números enteros positivos o negativos, desde el infinito negativo hasta el infinito. Tenga en cuenta que los enteros no deben incluir puntos decimales. Los enteros tienen el tipo `int`.

In [78]:
ano_actual = 2020

In [79]:
ano_actual

2020

In [80]:
type(ano_actual)

int

A diferencia de algunos otros lenguajes de programación, los enteros de Python pueden ser arbitrariamente grandes (o pequeños). No hay un valor más bajo o más alto para los enteros, y solo hay un tipo `int`.

In [81]:
numero_muy_largo = -23374038374832934334234317348343

In [82]:
numero_muy_largo

-23374038374832934334234317348343

In [83]:
type(numero_muy_largo)

int

#### Flotante o Decimal
Los flotantes (o números de punto flotante) son números con un punto decimal. No hay límites en el valor o el número de dígitos antes o después del punto decimal. Los números de punto flotante tienen el tipo `float`.

In [84]:
pi = 3.141592653589793238

In [85]:
pi

3.141592653589793

In [86]:
type(pi)

float

Ten en cuenta que un número entero se trata como un float si se escribe con un punto decimal, aunque la parte decimal del número es cero.

In [87]:
numero_1 = 3.0

In [88]:
numero_1

3.0

In [89]:
type(numero_1)

float

In [90]:
otro_numero = 4.

In [91]:
otro_numero

4.0

In [92]:
type(otro_numero)

float

Los números de punto flotante también se pueden escribir utilizando la notación científica con una "e" para indicar la potencia de 10.

In [93]:
un_centenar = 1e-2

In [94]:
un_centenar

0.01

In [95]:
type(un_centenar)

float

In [96]:
numero_mixto = 6.02214076e23

In [97]:
numero_mixto

6.02214076e+23

In [98]:
type(numero_mixto)

float

Puede convertir numeros flotantes a enteros y viceversa mediante las funciones `float` y `int`. La operación de convertir un tipo de valor en otro se denomina conversión.

In [99]:
float(ano_actual)

2020.0

In [100]:
float(numero_muy_largo)

-2.3374038374832935e+31

In [101]:
int(pi)

3

In [102]:
int(numero_mixto)

602214075999999987023872

Al realizar operaciones aritméticas, los enteros se convierten automáticamente en `float` si alguno de los operandos es un valor del tipo `float`. Además, el operador de división siempre devuelve un resultadoadel tipo `float`, incluso si ambos operandos son enteros. Utilice el operador `//` si desea que el resultado de la división sea `int`.

In [103]:
type(45 * 3.0)

float

In [104]:
type(45 * 3)

int

In [105]:
type(10/2)

float

In [106]:
type(10/3)

float

In [107]:
type(10//2)

int

#### Booleana
Los booleanos se representan con uno de 2 valores: `True` y `False`. Los booleanos tienen el tipo `boll`.

In [108]:
hoy_es_Sabado = True

In [109]:
hoy_es_Sabado

True

In [110]:
type(hoy_es_Sabado)

bool